### PR TITLE
Reject spans that create a circular loop

### PR DESCRIPTION
**Pull request by JamesNK**
## Description

Spans have a parent. A span is not allowed reference itself or one of its anscetors as a parent.

Data sent to the dashboard could contain an invalid circular reference. Rendering the bad data in the UI could cause errors. This PR adds validation logic when spans are received to reject the bad data.

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [x] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/6262)

### CODE REVIEW

File: src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs [hunk 1]
```diff
@@ -96,6 +101,31 @@ static string BuildFullName(OtlpSpan existingSpan)
         }
     }
 
+    private static bool HasCircularReference(OtlpSpan span)
```
Comment (mitchdenny) [inline src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs:104]: Is this called every single time a span is sent to the OTEL endpoint?

Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs:104]: Yes

File: src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs [hunk 2]
```diff
@@ -96,6 +101,31 @@ static string BuildFullName(OtlpSpan existingSpan)
...
+        {
+            return false;
+        }
+
+        // Walk up span ancestors to check there is no loop.
+        var stack = new List<OtlpSpan> { span };
+        var currentSpan = span;
+        while (currentSpan.GetParentSpan() is { } parentSpan)
+        {
+            if (stack.Contains(parentSpan))
```
Comment (drewnoakes) [inline src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs:117]: Is a `List<>` the right collection here? Seems like a `HashSet<>` would be a better fit, avoiding O(N) lookups.

Concerned about the space and time complexity of this change. As called out above, the method's called for every span, and it then walks the whole ancestry, and does an O(N) lookup within that. That's a lot of nested loops.

Comment (JamesNK) [reply src/Aspire.Dashboard/Otlp/Model/OtlpTrace.cs:122]: Changed to a keyed collection. Will adaptively switch to a dictionary when it exceeds a certain number of spans.

**Comment by JamesNK** [comment]
PR feedback addressed.

**Review by drewnoakes — APPROVED** [review]

**Pull request merged**